### PR TITLE
[DOC] Include guidelines for private functions in contributing docs

### PR DESCRIPTION
**Pull request by ymzayek**
- Addresses part of #3628

I just copied the example from the issue with @jeromedockes cited as the source at the bottom of the example (but it won't render in the html). IS it ok?

### CODE REVIEW

**Review by Remi-Gau — APPROVED** [review]
Good with me

**Review by bthirion — APPROVED** [review]
LGTM. Very useful, indeed !

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -351,6 +351,75 @@ This is also useful for writing unit tests.
 Writing small functions is not always possible, and we do not recommend trying to reorganize larger,
 but well-tested, older functions in the codebase, unless there is a strong reason to do so (e.g., when adding a new feature).
 
+.. _private_functions:
+
+Guidelines for Private Functions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In some cases when private functions are in a private module (filename beginning with an underscore),
+but are used outside of that file, we do not name them with a leading underscore. Example:
```
Comment (jeromedockes) [inline CONTRIBUTING.rst:360]: thanks! maybe this summary here could read something like:

we start a name with a leading underscore to indicate that it is an internal implementation detail, not to be accessed directly from outside, of the enclosing context: the parent module (for a submodule name), the module (for the name of a top-level function, class or global variable), or the class (for a method or attribute name). Moreover, modules explicitly declare their api through the `__all__` attribute, and any name not listed in `__all__` should not be accessed from outside the module

**Comment by Remi-Gau** [comment]
Sorry @jeromedockes I merged this too quickly: will take your comment in a separate issue, so we do not forget about it

**Pull request merged**